p4-16/psa/PSA.mdk

```diff
@@ -199,6 +202,41 @@ copy going to the control CPU to have a software-defined header
 (defined in the P4 program) that is different than any headers on the
 packet copies going to the Egress control block.
 
+A PSA implementation may choose to implement multiple class-of-service
+queues for each output port, but it may also choose to implement only
+1 queue for each output port.  The Ingress control block may choose to
+assign a value to the `ostd.class_of_service` field to change the
+packet's class of service to a value other than the default of 0.
+
+PSA only specifies how the Ingress control block can choose among
+multiple class-of-service queues for an output port, if more than one
+exists.  It does not mandate a scheduling policy among these queues.
+Something at least as flexible as weighted fair queuing, with an
+optional strict high priority queue, is recommended.
+
+The control plane API excerpt below is intended to be added as part of
+the P4Runtime API[^P4RuntimeAPI].
+
+```
+// The ClassOfServiceInfo message should be added to the "oneof"
+// inside of message "Entity".
+
+// ClassOfServiceInfo is only intended to be read.  Attempts to update
+// this entity have no effect, and should return an error status that
+// the entity is read only.
+
+message ClassOfServiceInfo {
```

Comment:
If this is just to retrieve switch capabilities information, we actually may not expose it in P4Runtime but in the Openconfig data models. For example, we plan on using Openconfig for port configuration (enable / disable ports, query port status, ...). IMO the COS information would belong there as well based on this proposal.

Also, is there no intention of exposing an API to configure scheduling and rate-limiting based on COS id? If this is considered out of scope of PSA, I'm not sure this document should include any discussion about COS runtime APIs.

Reply:
The intent is that it is only for retrieving capability information about the device, as the PSA currently allows that different PSA devices may implement a different number of class of services from each other.

I certainly think it would be useful if there were an API that all controllers could use to configure PSA devices that have multiple classes of service.  If you can get a proposal to pass Vladimir's variety of implementations he doesn't want to exclude, go for it :-)

Reply:
Would config information like the rate that timestamp_t values advance in a device also belong in an Openconfig data model, vs. the P4Runtime API?  I had earlier created some proposed P4Runtime API messages to retrieve that data from a device, too.

Reply:
@antoninbas @cc10512 I would recommend merging these changes in as written right now, and then do a later pass of editing to clean up control plane API issues, focused just on that topic.

Reply:
I'm fine with including the definition, however, we may want to say that these are guidelines that will be proposed to the P4Runtime WG. E.g.:
> The control plane API excerpt below is intended as a guideline to be included as part of the P4Runtime API[^P4RuntimeAPI] once discussed and approved.

Reply:
I added a note marked with "TBD" in the PSA spec for now, reminding us to revisit this part of the proposed control plane API.  If it is better to explain it differently than it is written up currently, I'm open to suggestions, but would like to get the rest of these changes merged in before that decision is made, if that is acceptable.